drivers/wifi/esp8266/wifi_esp8266.c

```diff
@@ -0,0 +1,1452 @@
+	}
+}
+
+static void on_cmd_sendok(struct net_buf **buf, u16_t len)
+{
+	foo_data.last_error = 0;
+
+	/* only signal semaphore if we are not waiting on a data send command */
+	k_sem_give(&foo_data.response_sem);
+}
+
+/* Handler: OK */
+static void on_cmd_sockok(struct net_buf **buf, u16_t len)
+{
+	foo_data.last_error = 0;
+
+	if (!foo_data.sending_data) {
+		k_sem_give(&foo_data.response_sem);
+	}
+}
+
+/* Handler: ERROR */
+static void on_cmd_sockerror(struct net_buf **buf, u16_t len)
+{
+	foo_data.last_error = -EIO;
+	k_sem_give(&foo_data.response_sem);
+}
+
+static void on_cmd_wifi_connected_resp(struct net_buf **buf, u16_t len)
+{
+	wifi_mgmt_raise_connect_result_event(foo_data.iface, 0);
+}
+
+static void on_cmd_wifi_disconnected_resp(struct net_buf **buf, u16_t len)
+{
+	wifi_mgmt_raise_disconnect_result_event(foo_data.iface, 0);
+}
+
+static int esp8266_get(sa_family_t family,
+			enum net_sock_type type,
+			enum net_ip_protocol ip_proto,
+			struct net_context **context)
+{
+	int i;
+	struct net_context *c = *context;
+
+	if (family != AF_INET) {
+		return -1;
+	}
+
+	k_sem_take(&sock_sem, K_FOREVER);
+	for (i = 0; i < ESP8266_MAX_CONNECTIONS && foo_data.sock_map & BIT(i); i++){}
+	if (i >= ESP8266_MAX_CONNECTIONS) {
+		k_sem_give(&sock_sem);
+		return 1;
+	}
+
+	foo_data.sock_map |= 1 << i;
+	c->offload_context = (void *)&foo_data.socket_data[i];
+	foo_data.socket_data[i].context = c;
+	foo_data.socket_data[i].family = family;
+	foo_data.socket_data[i].type = type;
+	foo_data.socket_data[i].ip_proto = ip_proto;
+	foo_data.socket_data[i].connected = 0;
+
+	k_sem_give(&sock_sem);
+	return 0;
+}
+
+static const char type_tcp[] = "TCP";
+static const char type_udp[] = "UDP";
+
+static int esp8266_connect(struct net_context *context,
+			const struct sockaddr *addr,
+			socklen_t addrlen,
+			net_context_connect_cb_t cb,
+			s32_t timeout,
+			void *user_data)
+{
+	char connect_msg[100];
+	int len;
+	const char *type;
+	struct socket_data *sock;
+	char addr_str[32];
+	int ret = -EFAULT;
+
+	if (!context || !addr) {
+		return -EINVAL;
+	}
+
+	sock = (struct socket_data *)context->offload_context;
+	if (!sock) {
+		LOG_ERR("Can't find socket info for ctx: %p\n", context);
+		return -EINVAL;
+	}
+
+#if defined (CONFIG_NET_IPV4)
+	if (addr->sa_family == AF_INET) {
+		net_ipaddr_copy(&net_sin(&sock->dst)->sin_addr,
+				&net_sin(addr)->sin_addr);
+		net_sin(&sock->dst)->sin_port = net_sin(addr)->sin_port;
+		sock->dst.sa_family = AF_INET;
+	} else
+#endif
+	{
+		return -EPFNOSUPPORT;
+	}
+
+	if (net_sin(&sock->dst)->sin_port < 0) {
+		LOG_ERR("invalid port: %d\n",
+			    net_sin(&sock->dst)->sin_port);
+		return -EINVAL;
+	}
+
+	if (net_context_get_type(context) == SOCK_STREAM) {
+		type = type_tcp;
+	} else {
+		type = type_udp;
+	}
+
+	if (type == type_tcp) {
+		net_addr_ntop(sock->dst.sa_family,
+			&net_sin(&sock->dst)->sin_addr,
+			addr_str, sizeof(addr_str));
+	len = sprintf(connect_msg, "AT+CIPSTART=%d,\"%s\",\"%s\",%d,7200",
```

Comment:
Use `snprintk()` here, no use of `sprintf()` as it does not check the size of the buffer. Make sure this is done through the whole file.